sklearn/metrics/base.py

```diff
@@ -131,3 +131,66 @@ def _average_binary_score(binary_metric, y_true, y_score, average,
         return np.average(score, weights=average_weight)
     else:
         return score
+
+
+def _average_multiclass_ovo_score(binary_metric, y_true, y_score, average):
+    """Uses the binary metric for one-vs-one multiclass classification,
+    where the score is computed according to the Hand & Till (2001) algorithm.
+
+    Parameters
+    ----------
+    y_true : array, shape = [n_samples]
+        True multiclass labels.
+        Assumes labels have been recoded to 0 to n_classes.
+
+    y_score : array, shape = [n_samples, n_classes]
+        Target scores corresponding to probability estimates of a sample
+        belonging to a particular class
+
+    average : 'macro' or 'weighted', default='macro'
+        ``'macro'``:
+            Calculate metrics for each label, and find their unweighted
+            mean. This does not take label imbalance into account. Classes
+            are assumed to be uniformly distributed.
+        ``'weighted'``:
+            Calculate metrics for each label, taking into account the a priori
+            distribution of the classes.
+
+    binary_metric : callable, the binary metric function to use.
+        Accepts the following as input
+            y_true_target : array, shape = [n_samples_target]
+                Some sub-array of y_true for a pair of classes designated
+                positive and negative in the one-vs-one scheme.
+            y_score_target : array, shape = [n_samples_target]
+                Scores corresponding to the probability estimates
+                of a sample belonging to the designated positive class label
+
+    Returns
+    -------
+    score : float
+        Average the sum of pairwise binary metric scores
+    """
+    n_labels = len(np.unique(y_true))
+    apriori_label_distribution = np.bincount(y_true) / float(y_true.size)
+    label_scores = np.zeros(n_labels)
+    for pos in range(n_labels):
+        for neg in range(pos + 1, n_labels):
+            ix = np.in1d(y_true.ravel(), [pos, neg])
+            y_true_filtered = y_true[ix.reshape(y_true.shape)]
+            y_score_filtered = y_score[ix]
+
+            # compute score with `pos` as the positive class
+            class_a = y_true_filtered == pos
+            # compute score with `neg` as the positive class
+            class_b = y_true_filtered == neg
+            score_class_a = binary_metric(
+                    class_a, y_score_filtered[:, pos])
+            score_class_b = binary_metric(
+                    class_b, y_score_filtered[:, neg])
+            binary_avg_score = (score_class_a + score_class_b) / 2.0
+            label_scores[pos] += binary_avg_score
+    if average == "weighted":
+        label_scores = np.multiply(apriori_label_distribution, label_scores)
+        return 2. * np.sum(label_scores) / (n_labels - 1)
```

Comment:
You're doubling here, but halving every contribution to label_scores 4 lines above.